### PR TITLE
Bug 1868551: Added Node Modal for BM OCS install

### PR DESCRIPTION
**Pull request by gnehapk**
- Added Node Modal (UXD- https://marvelapp.com/prototype/ai19e9g/screen/71440075)
- Changed type for size from string to number

Screenshot from 2020-08-11 13-34-59

Screenshot from 2020-08-13 12-32-10

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/create-local-volume-set.tsx [hunk 1]
```diff
@@ -52,16 +49,16 @@ export const CreateLocalVolumeSet: React.FC<CreateLocalVolumeSetProps> = ({ stat
...
         <Alert
           className="co-alert ceph-ocs-install__wizard-alert"
           variant="danger"
           title="Minimum Node Requirement"
           isInline
         >
           The OCS storage cluster require a minimum of 3 nodes for the intial deployment. Only{' '}
-          {nodesCnt} nodes match to the selected filters. Please adjust the filters to include more
-          nodes.
+          {state.filteredNodes.length} nodes match to the selected filters. Please adjust the
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/create-local-volume-set.tsx:60]:
```suggestion
          {state.filteredNodes.length} ${plularize('node', state.filteredNodes.length)} match to the selected filters. Please adjust the
```

File: frontend/packages/local-storage-operator-plugin/src/components/local-volume-set/local-volume-set-inner.tsx [hunk 2]
```diff
@@ -81,7 +81,7 @@ export const LocalVolumeSetInner: React.FC<LocalVolumeSetInnerProps> = ({
       <FormGroup label="Node Selector" fieldId="create-lvs--radio-group-node-selector">
         <div id="create-lvs-radio-group-node-selector">
           <Radio
-            label="All nodes"
+            label={`All nodes (${state.nodeNamesForLVS.length} nodes)`}
```
Comment (bipuladh) [inline frontend/packages/local-storage-operator-plugin/src/components/local-volume-set/local-volume-set-inner.tsx:84]: Use pluralize here as well.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx [hunk 3]
```diff
@@ -95,10 +113,40 @@ export const DiscoveryDonutChart: React.FC<DiscoveryDonutChartProps> = ({ state,
...
       <DiskListModal state={state} dispatch={dispatch} />
+      <NodeListModal state={state} dispatch={dispatch} />
     </div>
   );
 };
 
+const NodeListModal: React.FC<DiscoveryDonutChartProps> = ({ state, dispatch }) => {
+  const cancel = () => {
+    dispatch({ type: 'setShowNodeList', value: false });
+  };
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:124]:
```suggestion
  const cancel = () => dispatch({ type: 'setShowNodeList', value: false });
```

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx [hunk 4]
```diff
@@ -65,7 +74,16 @@ export const DiscoveryDonutChart: React.FC<DiscoveryDonutChartProps> = ({ state,
     <div className="ceph-ocs-install__chart-wrapper">
       <div className="ceph-ocs-install_capacity-header">Selected Capacity</div>
       <div className="ceph-ocs-install__stats">
-        <div>{pluralize(nodes.length, 'Node')}</div>
+        <div>
+          <Button
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:78]: Shouldn't this be disabled for 0 nodes?

Comment (gnehapk) [reply frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:78]: Yes, it should be. Will update.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx [hunk 5]
```diff
@@ -65,17 +74,34 @@ export const DiscoveryDonutChart: React.FC<DiscoveryDonutChartProps> = ({ state,
...
       <div className="ceph-ocs-install_capacity-header">Selected Capacity</div>
       <div className="ceph-ocs-install__stats">
-        <div>{pluralize(nodes.length, 'Node')}</div>
+        <div>
+          {state.filteredNodes.length ? (
+            <Button
+              component="a"
+              variant="link"
+              onClick={() => dispatch({ type: 'setShowNodeList', value: true })}
+              className="ceph-ocs-install__node-list-btn"
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:83]:
```suggestion
              variant="link"
              onClick={() => dispatch({ type: 'setShowNodeList', value: true })}
              className="ceph-ocs-install__node-list-btn"
```
Any specific reason to make it an "anchor" tag ?
If you are doing that you also need to have `href` .

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/attached-devices.scss [hunk 6]
```diff
@@ -64,7 +64,14 @@
   cursor: pointer;
 }
 
-.ceph-ocs-install__disks-modal {
+.ceph-ocs-install__node-list-btn {
+  padding-top: 0;
+  padding-left: 0;
+  padding-right: 0;
+  cursor: pointer;
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/attached-devices.scss:71]:
```suggestion
```
If you are making the variant as "link" in Button then you will get the cursor as "pointer". You should not need that to set explicitly.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/sc-node-list.tsx [hunk 7]
```diff
@@ -50,7 +50,7 @@ const getRows: GetRows = ({ componentProps, customData }) => {
     };
   });
 
-  if (!_.isEqual(filteredData, nodes)) {
+  if (setNodes && !_.isEqual(filteredData, nodes)) {
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/sc-node-list.tsx:53]: Why is this required?

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx [hunk 8]
```diff
@@ -65,17 +74,34 @@ export const DiscoveryDonutChart: React.FC<DiscoveryDonutChartProps> = ({ state,
     <div className="ceph-ocs-install__chart-wrapper">
       <div className="ceph-ocs-install_capacity-header">Selected Capacity</div>
       <div className="ceph-ocs-install__stats">
-        <div>{pluralize(nodes.length, 'Node')}</div>
+        <div>
+          {state.filteredNodes.length ? (
+            <Button
+              component="a"
+              variant="link"
+              onClick={() => dispatch({ type: 'setShowNodeList', value: true })}
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:82]:
```suggestion
              onClick={dispatch({ type: 'setShowNodeList', value: true })}
```

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx [hunk 9]
```diff
@@ -56,6 +59,12 @@ export const DiscoveryDonutChart: React.FC<DiscoveryDonutChartProps> = ({ state,
     dispatch,
   ]);
 
+  React.useEffect(() => {
+    const filteredNodes = new Set<string>();
+    state.filteredDiscoveries.forEach((discovery) => filteredNodes.add(discovery.node));
+    dispatch({ type: 'setFilteredNodes', value: Array.from(filteredNodes) });
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:65]: Array -> Set -> Array...? _.uniqBy

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/sc-node-list.tsx (hunk 7 above)
Comment (gnehapk) [reply frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/sc-node-list.tsx:53]: Because its used here too - https://github.com/openshift/console/pull/6289/files#diff-99347a73d43d6e64cba499ad4c8093edR147 and setNodes is not required.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx (hunk 9 above)
Comment (gnehapk) [reply frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:65]: Its just `Set -> Array` and its fine to use this way.

File: frontend/packages/ceph-storage-plugin/src/utils/install.ts [hunk 10]
```diff
@@ -37,7 +37,7 @@ export const filterSCWithoutNoProv = (sc: StorageClassResourceKind) =>
 export const getTotalDeviceCapacity = (list: Discoveries[]) => {
   const totalCapacity = list.reduce((res, device) => {
     if (device?.status?.state === AVAILABLE) {
-      const capacity = Number(convertToBaseValue(device.size));
+      const capacity = Number(convertToBaseValue(device.size.toString()));
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/utils/install.ts:40]: Size is already coming as `integer` value only so we can directly use it.
 `convertToBaseValue` will not be required.
```suggestion
      const capacity = device.size;
```

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx [hunk 11]
```diff
@@ -65,17 +74,34 @@ export const DiscoveryDonutChart: React.FC<DiscoveryDonutChartProps> = ({ state,
...
-            component="a"
-            variant="link"
-            onClick={() => dispatch({ type: 'setShowDiskList', value: true })}
-            className="ceph-ocs-install__disk-list-btn"
-          >
-            {pluralize(state.filteredDiscoveries.length, 'Disk')}
-          </Button>
+          {state.filteredDiscoveries.length ? (
+            <Button
+              component="a"
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:95]:
```suggestion
```
same as above

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx (hunk 8 above)
Comment (gnehapk) [reply frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx:82]: This will give you  - `Type 'void' is not assignable to type '(event: MouseEvent<HTMLButtonElement, MouseEvent>) => void'.ts(2322)` esLint error.

**Comment by gnehapk** [comment]
Depends on https://github.com/openshift/local-storage-operator/pull/135
/hold

**Review by bipuladh — APPROVED** [review]

**Comment by spadgett** [comment]
/hold

```
ERROR in /go/src/github.com/openshift/console/frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/donut-chart.tsx
(147,9): Type '{ kind: string; showTitle: false; ListComponent: FC<NodeTableProps>; hideLabelFilter: true; hideToolbar: true; customData: { filteredNodes: string[]; }; }' is not assignable to type 'IntrinsicAttributes & { ListComponent: ComponentType<any>; kind: string; helpText?: any; namespace?: string; filterLabel?: string; textFilter?: string; ... 15 more ...; hideColumnManagement?: boolean; } & { ...; }'.
  Property 'hideToolbar' does not exist on type 'IntrinsicAttributes & { ListComponent: ComponentType<any>; kind: string; helpText?: any; namespace?: string; filterLabel?: string; textFilter?: string; ... 15 more ...; hideColumnManagement?: boolean; } & { ...; }'.
```

**Comment by gnehapk** [comment]
/hold cancel 

since the eslint issue has been fixed.
